File: srv/src/server/src/server_s.cpp
#include "ros/ros.h"
#include "server/Addints.h"//包含头文件
/*
    服务器实现:
        1.包含头文件
        2.初始化 ROS 节点
        3.创建 ROS 句柄
        4.创建 服务 对象
        5.回调函数处理请求并产生响应
        6.由于请求有多个，需要调用 ros::spin()
*/
bool doNums(server::Addints::Request& request, server::Addints::Response& response)
{
    //处理请求
    int num1=request.num1;
    int num2=request.num2;
    ROS_INFO("收到的请求数据:num1=%d,num2=%d",num1,num2);
    //组织响应
    int sum=num1+num2;
    response.sum=sum;
    ROS_INFO("求和结果: sum=%d", response.sum);
    return true;
}

int main(int argc,char *argv[])
{
    setlocale(LC_ALL,"");//防止中文乱码所需的代码

    ros::init(argc,argv,"baojie");//初始化ROS节点
    ros::NodeHandle nh;//创建ROS句柄
    ros::ServiceServer server =nh.advertiseService("Jackson",doNums);//创建一个服务对象，Jackson为话题名，doNums为回调函数
    ROS_INFO("服务端启动！");
    ros::spin();//回调
    return 0;
}

File: srv/src/server/scripts/server_s.py
#! /usr/bin/env python
import rospy
from server.srv import Addints,AddintsRequest,AddintsResponse#导包
# 服务器端实现:
#         1.导包
#         2.初始化 ROS 节点
#         3.创建服务对象
#         4.回调函数处理请求并产生响应
#         5.spin 函数

def doNum(request):#回调函数【参数——封装了请求数据，返回值——响应数据】
    #1.解析提交的两个整数
    num1=request.num1
    num2=request.num2
    #2.求和
    sum=num1+num2
    #3.将结果封装进入响应对象
    response=AddintsResponse()
    response.sum=sum
    rospy.loginfo("服务器解析的数据num1=%d ,num2=%d ,结果:sum=%d ",num1,num2,sum)
    return response

if __name__=="__main__":
    rospy.init_node("XYY")#初始化ROS节点，XYY是自设定的节点名
    server=rospy.Service("JK",Addints,doNum)#创建服务器对象,JK为话题名，Addints为srv目录下的名称
    rospy.loginfo("服务器已经启动了!")
    #处理逻辑（回调函数）
    #spin（）
    rospy.spin()


